File: media/base/bit_reader_unittest.cc
// Copyright (c) 2012 The Chromium Authors. All rights reserved.
// Use of this source code is governed by a BSD-style license that can be
// found in the LICENSE file.

#include "media/base/bit_reader.h"

#include "testing/gtest/include/gtest/gtest.h"

namespace media {

TEST(BitReaderTest, EmptyStreamTest) {
  BitReader reader(NULL, 0);
  uint8 value8 = 0xff;

  ASSERT_FALSE(reader.HasMoreData());
  ASSERT_EQ(reader.Tell(), 0);
  ASSERT_TRUE(reader.ReadBits(0, &value8));
  ASSERT_TRUE(reader.SkipBits(0));
  ASSERT_EQ(reader.Tell(), 0);
  ASSERT_FALSE(reader.HasMoreData());
  ASSERT_FALSE(reader.ReadBits(1, &value8));
  ASSERT_FALSE(reader.SkipBits(1));
  ASSERT_EQ(value8, 0);
}

TEST(BitReaderTest, NormalOperationTest) {
  // 0101 0101 1001 1001 repeats 4 times
  uint8 buffer[] = {0x55, 0x99, 0x55, 0x99, 0x55, 0x99, 0x55, 0x99};
  BitReader reader(buffer, 6);  // Initialize with 6 bytes only
  uint8 value8;
  uint64 value64;

  ASSERT_TRUE(reader.HasMoreData());
  ASSERT_EQ(reader.Tell(), 0);
  ASSERT_TRUE(reader.ReadBits(1, &value8));
  ASSERT_EQ(value8, 0);
  ASSERT_TRUE(reader.ReadBits(8, &value8));
  ASSERT_EQ(value8, 0xab);  // 1010 1011
  ASSERT_EQ(reader.Tell(), 9);
  ASSERT_TRUE(reader.HasMoreData());
  ASSERT_TRUE(reader.SkipBits(7));
  ASSERT_EQ(reader.Tell(), 16);
  ASSERT_TRUE(reader.ReadBits(32, &value64));
  ASSERT_EQ(value64, 0x55995599u);
  ASSERT_EQ(reader.Tell(), 48);
  ASSERT_FALSE(reader.HasMoreData());
  ASSERT_FALSE(reader.SkipBits(1));
  ASSERT_FALSE(reader.ReadBits(1, &value8));
  ASSERT_TRUE(reader.SkipBits(0));
  value8 = 0xff;
  ASSERT_TRUE(reader.ReadBits(0, &value8));
  ASSERT_EQ(value8, 0);

  reader.Initialize(buffer, 8);
  ASSERT_TRUE(reader.ReadBits(64, &value64));
  EXPECT_EQ(value64, 0x5599559955995599ull);
  EXPECT_FALSE(reader.HasMoreData());
  EXPECT_EQ(reader.Tell(), 64);
  EXPECT_FALSE(reader.ReadBits(1, &value8));
  EXPECT_FALSE(reader.SkipBits(1));
  EXPECT_TRUE(reader.ReadBits(0, &value8));
  EXPECT_TRUE(reader.SkipBits(0));
}

TEST(BitReaderTest, LongSkipTest) {
  uint8 buffer[] = {
    0x12, 0x34, 0x56, 0x78, 0x12, 0x34, 0x56, 0x78,  // 64 * 1
    0x12, 0x34, 0x56, 0x78, 0x12, 0x34, 0x56, 0x78,  // 64 * 2
    0x12, 0x34, 0x56, 0x78, 0x12, 0x34, 0x56, 0x78,  // 64 * 3
    0x12, 0x34, 0x56, 0x78, 0x12, 0x34, 0x56, 0x78,  // 64 * 4
    0x12, 0x34, 0x56, 0x78, 0x12, 0x34, 0x56, 0x78,  // 64 * 5
    0x12, 0x34, 0x56, 0x78, 0x12, 0x34, 0x56, 0x78,  // 64 * 6
    0x12, 0x34, 0x56, 0x78, 0x12, 0x34, 0x56, 0x78,  // 64 * 7
    0x12, 0x34, 0x56, 0x78, 0x12, 0x34, 0x56, 0x78,  // 64 * 8
    0x87, 0x65
  };
  BitReader reader(buffer, sizeof(buffer));
  uint8 value8;

  EXPECT_TRUE(reader.SkipBits(64 * 8 + 8));
  EXPECT_EQ(reader.Tell(), 64 * 8 + 8);
  EXPECT_TRUE(reader.ReadBits(8, &value8));
  EXPECT_EQ(value8, 0x65);
  EXPECT_EQ(reader.Tell(), 64 * 8 + 16);
  EXPECT_FALSE(reader.HasMoreData());
  EXPECT_EQ(reader.NumBitsLeft(), 0);
}

TEST(BitReaderTest, ReadBeyondEndTest) {
  uint8 buffer[] = {0x12};
  BitReader reader(buffer, sizeof(buffer));
  uint8 value8;

  EXPECT_TRUE(reader.SkipBits(4));
  EXPECT_FALSE(reader.ReadBits(5, &value8));
  EXPECT_FALSE(reader.ReadBits(1, &value8));
  EXPECT_FALSE(reader.SkipBits(1));
  EXPECT_TRUE(reader.ReadBits(0, &value8));
  EXPECT_TRUE(reader.SkipBits(0));

  reader.Initialize(buffer, sizeof(buffer));
  EXPECT_TRUE(reader.SkipBits(4));
  EXPECT_FALSE(reader.SkipBits(5));
  EXPECT_FALSE(reader.ReadBits(1, &value8));
  EXPECT_FALSE(reader.SkipBits(1));
  EXPECT_TRUE(reader.ReadBits(0, &value8));
  EXPECT_TRUE(reader.SkipBits(0));
}

}  // namespace media

File: media/formats/mpeg/bit_reader_unittest.rs
// Copyright 2026 The Chromium Authors
// Use of this source code is governed by a BSD-style license that can be
// found in the LICENSE file.

chromium::import! {
    "//media/formats:safe_mpeg_audio_parsers";
}

use rust_gtest_interop::prelude::*;
use safe_mpeg_audio_parsers::bit_reader::SimpleBitReader;

#[gtest(BitReaderTest, NormalOperation)]
fn test_normal_operation() {
    // 0101 0101 1001 1001 repeats 4 times
    let buffer = [0x55, 0x99, 0x55, 0x99, 0x55, 0x99, 0x55, 0x99];
    let mut reader1 = SimpleBitReader::new(&buffer[..6]); // 6 bytes only

    expect_eq!(reader1.read_bits(1), Some(0));
    expect_eq!(reader1.read_bits(8), Some(0xab)); // 1010 1011
    expect_true!(reader1.read_bits(7).is_some());
    expect_eq!(reader1.read_bits(32), Some(0x55995599));
    expect_eq!(reader1.read_bits(1), None); // out of bounds!
    expect_eq!(reader1.read_bits(0), Some(0));

    let mut reader2 = SimpleBitReader::new(&buffer);
    expect_eq!(reader2.read_bits(32), Some(0x55995599));
    expect_eq!(reader2.read_bits(32), Some(0x55995599));
    expect_eq!(reader2.read_bits(1), None);
    expect_eq!(reader2.read_bits(0), Some(0));
}

#[gtest(BitReaderTest, ReadBeyondEnd)]
fn test_read_beyond_end() {
    let buffer = [0x12];
    let mut reader = SimpleBitReader::new(&buffer);

    expect_eq!(reader.read_bits(4), Some(0x01)); // 0001
    expect_eq!(reader.read_bits(5), None);
    expect_eq!(reader.read_bits(1), None);
    expect_eq!(reader.read_bits(0), Some(0));
}

#[gtest(BitReaderTest, SkipBits)]
fn test_skip_bits() {
    let buffer = [0x0a, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15];
    let mut reader = SimpleBitReader::new(&buffer);

    expect_true!(reader.skip_bits(2).is_some());
    expect_eq!(reader.read_bits(3), Some(1));
    expect_true!(reader.skip_bits(11).is_some());
    expect_eq!(reader.read_bits(8), Some(3));
    expect_true!(reader.skip_bits(76).is_some());
    expect_eq!(reader.read_bits(4), Some(13));
    expect_true!(reader.skip_bits(100).is_none());
    expect_true!(reader.skip_bits(usize::MAX).is_none());
    expect_true!(reader.skip_bits(0).is_some());
    expect_true!(reader.skip_bits(1).is_none());
}

#[gtest(BitReaderTest, BitsRead)]
fn test_bits_read() {
    let buffer = [0x0a, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15];
    let mut reader = SimpleBitReader::new(&buffer);

    expect_true!(reader.skip_bits(2).is_some());
    expect_eq!(reader.bits_read(), 2);
    expect_eq!(reader.read_bits(3), Some(1));
    expect_eq!(reader.bits_read(), 5);
    expect_eq!(reader.read_bits(1), Some(0)); // Read Flag
    expect_eq!(reader.bits_read(), 6);
    expect_true!(reader.skip_bits(76).is_some());
    expect_eq!(reader.bits_read(), 82);
}

#[gtest(BitReaderTest, SlowPathRefill)]
fn test_slow_path_refill() {
    // 1-byte buffer (< 4 bytes)
    let mut reader1 = SimpleBitReader::new(&[0xab]);
    expect_eq!(reader1.read_bits(4), Some(0xa));
    expect_eq!(reader1.read_bits(4), Some(0xb));
    expect_eq!(reader1.read_bits(1), None);

    // 2-byte buffer (< 4 bytes)
    let mut reader2 = SimpleBitReader::new(&[0x12, 0x34]);
    expect_eq!(reader2.read_bits(16), Some(0x1234));
    expect_eq!(reader2.read_bits(1), None);

    // 3-byte buffer (< 4 bytes)
    let mut reader3 = SimpleBitReader::new(&[0x55, 0x66, 0x77]);
    expect_eq!(reader3.read_bits(24), Some(0x556677));
    expect_eq!(reader3.read_bits(1), None);
}
